be/src/runtime/global_dict/decoder.cpp

```diff
@@ -16,52 +16,108 @@
 
 #include <utility>
 
+#include "column/array_column.h"
 #include "column/column_builder.h"
+#include "column/const_column.h"
+#include "column/datum.h"
+#include "column/nullable_column.h"
 #include "column/type_traits.h"
+#include "column/vectorized_fwd.h"
 #include "gutil/casts.h"
 #include "runtime/global_dict/config.h"
 #include "runtime/global_dict/types.h"
+#include "runtime/types.h"
 
 namespace starrocks {
 
-template <LogicalType type, typename Dict, LogicalType result_type>
+template <typename Dict>
 class GlobalDictDecoderBase : public GlobalDictDecoder {
 public:
-    using FieldType = RunTimeCppType<type>;
-    using ResultColumnType = RunTimeColumnType<result_type>;
-    using ColumnType = RunTimeColumnType<type>;
+    using DictCppType = RunTimeCppType<LowCardDictType>;
+    using DictColumnType = RunTimeColumnType<LowCardDictType>;
+    using StringColumnType = RunTimeColumnType<TYPE_VARCHAR>;
 
     GlobalDictDecoderBase(Dict dict) : _dict(std::move(dict)) {}
 
-    Status decode(Column* in, Column* out) override;
+    Status decode_string(Column* in, Column* out) override;
+
+    Status decode_array(Column* in, Column* out) override;
 
 private:
     Dict _dict;
 };
 
-template <LogicalType type, typename Dict, LogicalType result_type>
-Status GlobalDictDecoderBase<type, Dict, result_type>::decode(Column* in, Column* out) {
+template <typename Dict>
+Status GlobalDictDecoderBase<Dict>::decode_array(Column* in, Column* out) {
+    DCHECK(in != nullptr);
+    DCHECK(out != nullptr);
+
+    TypeDescriptor stringType;
+    stringType.type = TYPE_VARCHAR;
+    if (in->is_constant()) {
+        auto* const_column = down_cast<ConstColumn*>(in);
+        auto* in_array = down_cast<ArrayColumn*>(const_column->data_column().get());
+        auto in_element = in_array->elements_column();
+        auto in_offsets = in_array->offsets_column();
+
+        auto* out_array = down_cast<ArrayColumn*>(out);
+        auto out_element = out_array->elements_column();
+        auto out_offsets = out_array->offsets_column();
+        out_offsets->swap_column(*in_offsets);
+        return decode_string(in_element.get(), out_element.get());
```

Comment:
For nullable arrays, a slice in a partial null value is an illegal value, and I'm not sure they're safe.

Reply:
logical value:
[null, "key2", "key3"]
physical value:
nullcolumn: [1, 0, 0]
arraycolumn: ["key1", "key2", "key3"]

Reply:
don't worry, if string column has any value in actual, it's should has dict-id?